src/importexport/lyricsexport/internal/lrcwriter.cpp

```diff
@@ -0,0 +1,236 @@
+ */
+
+#include "lrcwriter.h"
+
+#include <QBuffer>
+
+#include "io/file.h"
+#include "types/ret.h"
+
+#include "engraving/dom/lyrics.h"
+#include "engraving/dom/masterscore.h"
+#include "engraving/dom/repeatlist.h"
+
+using namespace Qt::Literals;
+
+using namespace muse;
+using namespace muse::io;
+using namespace mu::engraving;
+using namespace mu::project;
+using namespace mu::iex::lrcexport;
+
+std::vector<INotationWriter::UnitType> LRCWriter::supportedUnitTypes() const
+{
+    return { UnitType::PER_PART };
+}
+
+bool LRCWriter::supportsUnitType(UnitType ut) const { return ut == UnitType::PER_PART; }
+
+muse::Ret LRCWriter::write(notation::INotationPtr notation, muse::io::IODevice& device, const Options&)
+{
+    Score* score = notation->elements()->msScore();
+    bool enhancedLrc = configuration()->lrcUseEnhancedFormat();
+
+    return doWrite(score, &device, enhancedLrc);
+}
+
+bool LRCWriter::writeScore(mu::engraving::Score* score, const muse::io::path_t& path, bool enhancedLrc)
+{
+    File f(path);
+    if (!f.open(IODevice::WriteOnly)) {
+        return false;
+    }
+
+    bool res = doWrite(score, &f, enhancedLrc) && !f.hasError();
+    f.close();
+
+    return res;
+}
+
+muse::Ret LRCWriter::writeList(const notation::INotationPtrList&, muse::io::IODevice&, const Options&)
+{
+    return make_ret(Ret::Code::NotSupported);
+}
+
+void LRCWriter::writeMetadata(muse::io::IODevice* device, const engraving::Score* score) const
+{
+    QString metadata;
+
+    // Title
+    const QString title = score->metaTag(u"workTitle").toQString();
+    if (!title.isEmpty()) {
+        metadata += u"[ti:%1]\n"_s.arg(title);
+    }
+
+    // Composer/Artist
+    const QString artist = score->metaTag(u"composer").toQString();
+    if (!artist.isEmpty()) {
+        metadata += u"[ar:%1]\n"_s.arg(artist);
+    }
+
+    if (!metadata.isEmpty()) {
+        device->write(metadata.toUtf8());
+    }
+}
+
+muse::Ret LRCWriter::doWrite(mu::engraving::Score* score, muse::io::IODevice* device, bool enhancedLrc)
+{
+    writeMetadata(device, score);
+
+    const auto lyrics = collectLyrics(score);
+
+    // Write lyrics
+    for (const auto& [timestamp, text] : lyrics) {
+        if (enhancedLrc) {
+            // As there should only be words we replace spaces by "-"
+            QString lyricsText = text;
+            lyricsText.replace(QRegularExpression("\\s"), "-");
+            lyricsText.replace(u'\u00A0', u'-');
+
+            device->write(QString("[%1] <%1> %2\n").arg(formatTimestamp(timestamp), lyricsText).toUtf8());
+        } else {
+            device->write(QString("[%1]%2\n").arg(formatTimestamp(timestamp), text).toUtf8());
+        }
+    }
+
+    return make_ok();
+}
+
+std::map<double, QString> LRCWriter::collectLyrics(const mu::engraving::Score* score)
+{
+    std::map<double, QString> lyrics;
+    const RepeatList& repeats = score->repeatList();
+
+    staff_idx_t lyricsStaff;
+    voice_idx_t lyricsVoice;
```

Comment:
Maybe we can improve it even further: 
* would it be a possibility to use `track_idx_t` instead of `staff_idx_t` and `voice_idx_t`? Keep in mind the following invariant: `track == staff * VOICES + voice`.
* keep in mind that the elements in `seg->elist()` appear in order of track. The first four tracks are the first staff, the second four are the second staff, etc.. That info might allow for more optimisations.